enhancements/machine-api/machine-pools-in-cluster.md

```diff
@@ -0,0 +1,240 @@
+authors:
+  - "@dgoodwin"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@JoelSpeed"
+  - "@enxebre"
+  - "@sdodson"
+  - "@staebler"
+approvers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@JoelSpeed"
+  - "@enxebre"
+  - "@sdodson"
+creation-date: 2020-09-21
+last-updated: 2020-09-21
+status: provisional
+---
+
+# AZ Balanced MachinePools In-Cluster
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today we expose a useful construct in the Installer's InstallConfig known as MachinePool. When used the installer will query the available AZs for the account, and generate the relevant MachineSets to spread the requested number of machine replicas out over the AZs.
+
+This enhancement proposes bringing this functionality into the cluster itself for use by customers on an on-going basis.
+
+
+## Motivation
+
+The Installer MachinePool is a useful construct that is only available at install time. Post-install customers must manage this spread themselves.
+
+[OpenShift Hive](https://github.com/openshift/hive) also supports day 2 MachinePools using the same code from the Installer. However because Hive must install and maintain MachineSets for many versions of OCP, this code re-use between the Installer has generated a number of difficult and messy bugs when things change in the Installer. Bringing this feature in-cluster would also give Hive a stable API to interact with instead of fighting with one version of vendored Go code to be used across many OCP versions and a significant number of logic forks based on OCP version.
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+  * Provide customers with an API to automatically spread machines out over the AZs in their account.
+  * Provide OpenShift Hive with a stable API to manage MachineSets spread over AZs day 2.
+
+### Non-Goals
+
+  * No linkup with the MachineDeployment concept for rolling updates of Machines.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As an OCP user, I would like to be able to specify min/max machine replicas and
+have those machines automatically spread out amongst all Availability Zones
+that are available in my account.
+
+#### Story 2
+
+As an OpenShift Hive developer, I would like a stable API for spreading
+machines across Availability Zones. This would allow me to sync an API
+in-cluster rather than use one version of vendored code from the installer to
+generate each MachineSet to be reconciled.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementation would require a new CRD and controller aligned hopefully with a pre-existing operator in OCP.
```

Comment:
There is upstream discussion about adding this first part into MachineSets (https://github.com/kubernetes-sigs/cluster-api/issues/3358), do we want to explore this as an alternative before we make a decision on implementing a new CRD?

Reply:
I would 100% prefer this, it's dramatically simpler. If there's a chance, that would be great.

Reply:
Need to walk back my comment above, after discussing with my team this would not help us with either use case really as MachineSet still requires substantial and non-trivial delicate settings that vary from OpenShift release to release. This makes it difficult for customers to just add their own, and for Hive to add them for customers and stay in sync with what the product expects. 

Examples: AMI ID, many fields with the infra ID specified in the name (label selectors, iamInstanceProfile, security group filters, subnet, tags), userDataSecret (name changed twice during the 4.6 cycle, once to new and then back, nearly caused 4.6 blocker because OSD could not install GCP as a result this week).

We still think the best approach here is to make an API to allow customers to easily add new pools of machines.

Reply:
Not sure if it helps or not, but we added a number of default options for MachineSets in the 4.6 cycle which has drastically reduced the number of options that have to be specified (as the defaulting webhook will fill them in on create anyway).

My concern over creating an extra CRD that creates MachineSets is that it will still need to have these kinds of options wouldn't it? Eg what if a customer wants to change things such as the subnet? Or will this just be a convenience with a limited set of options?

Reply:
Could the default hook cover everything I mentioned above? In conjunction with spreading over AZs, that might be workable. I will run this past the team.

No the machine pool API would not expose those settings, it's just meant to be a dead simple interface as it is today in hive and the InstallConfig. If you need control over subnets etc, you can go straight to MachineSet.